### PR TITLE
Metaclient - Leader election - leader change event notification 

### PR DESCRIPTION
**Pull request by xyuanlu**
### Issues

- [X] My PR addresses the following Helix issues and references them in the PR description:

 #2237

### Description

- [X] Here are some details about my PR, including screenshots of any UI changes:

This pr introduces leader election event listener.

### Tests

- [X] The following tests are written for this issue:

(List the names of added unit/integration tests)

- The following is the result of the "mvn test" command on the appropriate module:

TestLeaderElection.testLeadershipListener, TestLeaderElection.testRelinquishLeadership

### Changes that Break Backward Compatibility (Optional)

- My PR contains changes that break backward compatibility or previous assumptions for certain methods or API. They include:

(Consider including all behavior changes for public methods or API. Also include these changes in merge description so that other developers are aware of these changes. This allows them to make relevant code changes in feature branches accounting for the new method/API behavior.)

### Documentation (Optional)

- In case of new functionality, my PR adds documentation in the following wiki page:

(Link the GitHub wiki you added)

### Commits

- My commits all reference appropriate Apache Helix GitHub issues in their subject lines. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Code Quality

- My diff has been formatted using helix-style.xml 
(helix-style-intellij.xml if IntelliJ IDE is used)

### CODE REVIEW

File: meta-client/src/main/java/org/apache/helix/metaclient/recipes/leaderelection/LeaderElectionClient.java [hunk 1]
```diff
@@ -288,7 +289,8 @@ public List<String> getParticipants(String leaderPath) {
    * @return A boolean value indicating if registration is success.
    */
   public boolean subscribeLeadershipChanges(String leaderPath, LeaderElectionListenerInterface listener) {
-    //TODO: add converter class for LeaderElectionListenerInterface
+    _metaClient.subscribeDataChange(leaderPath + "/LEADER", new LeaderElectionListenerInterfaceAdapter(listener),
+        false);
```
Comment (junkaixue) [inline meta-client/src/main/java/org/apache/helix/metaclient/recipes/leaderelection/LeaderElectionClient.java:293]: Question: does this mean this is non-persist watcher or persist watcher?

If it is persist watch, if the ephemeral node is gone, then is the watcher still valid and receive new leader acquired notification?

File: meta-client/src/main/java/org/apache/helix/metaclient/recipes/leaderelection/LeaderElectionListenerInterface.java [hunk 2]
```diff
@@ -24,9 +24,16 @@
  * leader node is deleted.
  */
 public interface LeaderElectionListenerInterface {
+  enum ChangeType {
+    NEW_LEADER_ELECTED,
+    LEADER_GONE
```
Comment (junkaixue) [inline meta-client/src/main/java/org/apache/helix/metaclient/recipes/leaderelection/LeaderElectionListenerInterface.java:29]: Would suggest reword it:

LEADER_ACQUIRED
LEADER_LOST

File: meta-client/src/main/java/org/apache/helix/metaclient/recipes/leaderelection/LeaderElectionClient.java (hunk 1 above)
Comment (xyuanlu) [reply meta-client/src/main/java/org/apache/helix/metaclient/recipes/leaderelection/LeaderElectionClient.java:293]: It is persist watcher. If the ephemeral node is gone, then is the watcher still valid and receive new leader acquired notification.

File: meta-client/src/main/java/org/apache/helix/metaclient/recipes/leaderelection/LeaderElectionListenerInterface.java (hunk 2 above)
Comment (xyuanlu) [reply meta-client/src/main/java/org/apache/helix/metaclient/recipes/leaderelection/LeaderElectionListenerInterface.java:29]: Client can opt to listener on leader election changes and not essentially participant on leader election group. I feel like 'LEADER_ACQUIRED' is a bit confusing as it indicates the current client acquired/lost leadership

File: meta-client/src/main/java/org/apache/helix/metaclient/recipes/leaderelection/LeaderElectionClient.java (hunk 1 above)
Comment (junkaixue) [reply meta-client/src/main/java/org/apache/helix/metaclient/recipes/leaderelection/LeaderElectionClient.java:293]: Let's not hard code "LEADER"

**Review by junkaixue — APPROVED** [review]
Overall, lgtm.

**Comment by xyuanlu** [comment]
This change is ready to be merged. Approved by @junkaixue 
Commit message: Metaclient - Leader election - leader change event notification

**Pull request merged**